### PR TITLE
IGNITE-23413 Create RebalanceMinimumRequiredTimeProvider interface and its implementations

### PR DESCRIPTION
**Pull request by ibessonov**
https://issues.apache.org/jira/browse/IGNITE-23413

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/rebalance/RebalanceMinimumRequiredTimeProviderImpl.java [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+            int zonePartitions = zoneDescriptors.lastEntry().getValue().partitions();
+
+            Long pendingChangeTriggerKey = pendingChangeTriggerRevisions.get(tableId);
+
+            // +-1 here ir required for 2 reasons:
+            // - we need timestamp right before deletion, if zone is deleted, thus we must subtract 1;
+            // - we need a "metaStorageSafeTime" if zone is not deleted, without any subtractions.
+            long latestTimestamp = zoneDeletionTimestamps.getOrDefault(zoneId, metaStorageSafeTime + 1) - 1;
+
+            long zoneRevision = pendingChangeTriggerKey == null
```
Comment (sanpwc) [inline modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/rebalance/RebalanceMinimumRequiredTimeProviderImpl.java:106]: Let's say that zone was created on revision 10, and updated on revision 12. pendingChangeTriggerKey is null.
What will be the value of 'zoneRevision'?

File: modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/rebalance/RebalanceMinimumRequiredTimeProviderImpl.java [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+                    ? zoneDescriptors.firstEntry().getValue().updateToken()
+                    : pendingChangeTriggerKey;
+
+            NavigableMap<Long, CatalogZoneDescriptor> map = allZonesByRevision.get(zoneId);
+            Map.Entry<Long, CatalogZoneDescriptor> zone = map.floorEntry(zoneRevision);
+            long timestamp = metaStorageManager.timestampByRevisionLocally(zone.getValue().updateToken()).longValue();
+
+            timestamp = ceilTime(zoneDescriptors, timestamp, latestTimestamp);
```
Comment (sanpwc) [inline modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/rebalance/RebalanceMinimumRequiredTimeProviderImpl.java:114]: Hmm, Let's say that we have

- metaStorageSafeTime = t100.
- table with stable assignments only and thus pendingChangeTriggerKey == null.
- zone that was created at revision 10 (t10) and updated on revisions 20(t20) and 30 (t30).
- What will be the value of `minTimestamp = min(minTimestamp, timestamp);`?

File: modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/rebalance/RebalanceMinimumRequiredTimeProviderImpl.java (hunk 1 above)
Comment (ibessonov) [reply modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/rebalance/RebalanceMinimumRequiredTimeProviderImpl.java:106]: I should add such test and assert what we want to see, if it doesn't exist yer.
My assumption is - we should be able see the earliest catalog descriptor for a zone.

File: modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/rebalance/RebalanceMinimumRequiredTimeProviderImpl.java (hunk 2 above)
Comment (ibessonov) [reply modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/rebalance/RebalanceMinimumRequiredTimeProviderImpl.java:115]: Updated the code according to this comment. Now I distinguish between zone updates that do lead or don't lead to reassignment.

**Review by sanpwc — APPROVED** [review]

**Pull request merged**